src/Parallel/Algorithm.hpp

```diff
@@ -354,9 +358,15 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
           << ", and the halt flag is: " << halt_algorithm_until_next_phase_);
     }
     // set terminate to true if there are no actions in this PDAL
+
     set_terminate(number_of_actions_in_phase(next_phase) == 0);
+    phase_bookmarks_[phase_] = algorithm_step_;
```

Comment:
I'm confused.  This sets `phase_bookmarks_[phase_] = algorithm_step_` on start of the phase, but I expected this to be set at the _end_ of the phase so we can restart there. Currently, the below `if-else` will always take the `if` branch, which is also a no-op. What am I missing that makes this code magically work the way the PR description says it should?

Reply:
Perhaps you've overlooked the `phase_ = next_phase;` on the next line?

The point is that we don't have any particular code that runs at the end of phases, but we have this that runs at the beginning of every phase, or, equivalently, following every 'previous' phase. The line this comment is on (363) sets the bookmark for the previous phase (before we update `phase_`), then after we update `phase_` to the current phase, we check if we've previously set a bookmark (equivalent to checking if we have previously left the phase we are now entering). Then, if we have a bookmark, we start there. Otherwise, we start from the beginning.

I'll add code comments to that effect, unless now I'm missing something and there's a bug you can see in my reasoning.

Reply:
ah yep, that makes sense! I think this is probably the safest way of doing it, but I agree that some comments making that clear would be useful since it's fairly subtle :)